rfcs/53-generalized-hierarchy.md

```diff
@@ -0,0 +1,273 @@
+To signal changes in the hierachy, a `HierarchyEvent` event type is registered
+and any changes made to the hierarchy is published as an event.
+
+## Implementation strategy
+The core of this proposal is the `Hierarchy` component, which roughly looks like
+this:
+
+```rust
+#[derive(Component)]
+pub struct Hierarchy {
+  depth: u16,
+  parent: Option<Entity>,
+  prev: Option<Entity>,
+  next: Option<Entity>
+  first_child: Option<Entity>,
+}
+```
+
+This type has no public constructors and has no public APIs that allow for
+mutation of any of the internal data, only getters. The consistency between
+multiple instances of the component is maintained by a set of dedicated hierachy
+management commands. The following should be added as extensions to
+EntityCommands:
+
+  - `set_parent(Option<Entity>)` - sets the parent of a given entity.
+      - If `Hierarchy` is on neither parent or child, it will be added and both
+        will be atomically updated to reflect this change.
+      - If the parent already had children, the new child will be prepended at
+        the start of the linked list, and the sibling pointers of both will be
+        updated.
+      - If the child already had a parent, that previous parent will be updated
+        too.
+      - A HierarchyEvent is added to the queue to reflect this update.
+  - `move_to_first(Entity)` - Moves one of the children to the front.
+  - `move_to_last(Entity)` - Moves one of the children to the end.
+  - `remove_all_children()` - Disconnects all children from a parent. All
+    children become roots.
+      - A HierarchyEvent is added to the queue to reflect this update.
+
+All of these commands will also update the four marker components appropriately.
+Optionally, we could also mark these components as changed (even though they're
+ZSTs) to leverage change detection.
+
+One optional extension to this design is to create dedicated `HierarchyQuery<Q,
+F>`, which is a custom SystemParam that wraps multiple Query objects to simplify
+querying for items in the hierarchy:
+
+ - `iter_descendants(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does a depth-first traversal over the hierarchy
+   starting at some entity.
+ - `iter_ancestors(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This bubbles up queries from the bottom to the top
+   starting at some entity.
+
+As these operations are handled at any command buffer flush. The hierarchy
+maintenance system can be removed.
+
+## Benefits
+The biggest benefit to this new design is guarenteed global consistency. The
+hierarchy cannot be arbitrarily mutated between synchronization points.
+
+The second benefit is that any update to the hierarchy is "atomic" and is `O(1)`
+for the majority of the operations handled.
+
+## Drawbacks
+There are a notable number of drawbacks to this design:
+
+ - Counting children is now `O(n)`. Finding and editing children order is still
```

Comment:
It's a) not clear when we want to count children and b) not clear to me that the number of children will ever be greater than say 10 in realistic use cases.

Reply:
I can think of quite a few cases where 10 is reasonable and realistic, but this is no longer an issue as I switched the design up to keep `Parent`/`Children`.